### PR TITLE
Added client side logout

### PR DESCRIPTION
**Pull request by vassdeniss**
Added basic client side logout to the app, since it's been sitting there for some time.
Still needs to be able to delete the session token from the backend, but I couldn't manage to achieve that.

Edit: sessions are also now deleted on the backend

### CODE REVIEW

**Review by viktorpts — CHANGES_REQUESTED** [review]
The logout function is currently designed like a view (receiving context and redirecting), while its location suggests it's a service function. The files in `./src/data/` are supposed to only hold request logic without higher level operations. Additionally, the function contradicts Single Responsibility, since it both performs a request, operates with storage and controls the view.

I suggest only the request and storage operations remain here (similar to the other auth functions) while the view control is moved to another file in `./src/views/`

Another problem is the lack of error handling - if any of the async operations fails, the removal of local storage will not happen and the user's session may be soft-locked.

**Comment by vassdeniss** [comment]
Is it alright for a view to only really call the logout logic without displaying anything? If I understood the idea properly of course.

But yes otherwise It did felt wrong using the context in the `auth.js` file, that I will correct.

**Comment by vassdeniss** [comment]
I believe the new commit addresses SR and error handling.

File: src/data/auth.js [hunk 1]
```diff
@@ -26,14 +27,10 @@ export async function login(username, password) {
...
 
-export async function logout(ctx, next) {
-    const user = getUserData();
-    const [session] = (await get(endpoints.sessionByToken(user.sessionToken)))
+export async function logout(sessionToken) {
+    const [session] = (await get(endpoints.sessionByToken(sessionToken)))
         .results;
 
     await del(endpoints.sessionById(session.objectId));
     clearUserData();
```
Comment (viktorpts) [inline src/data/auth.js]: This line must execute unconditionally (currently if the async operations fail, this code wont be reached)

**Review by viktorpts — CHANGES_REQUESTED** [review]

File: src/data/auth.js (hunk 1 above)
Comment (vassdeniss) [reply src/data/auth.js]: If I understood correctly, why should we clear the local storage if we fail on the backend? Shouldn't we only want to clear both and only both, not individually?

Comment (viktorpts) [reply src/data/auth.js]: If the token isn't valid (malformed, or automatically expired on the backend) the requests will never go through and the storage wont be cleared, preventing all other functions of the app from working.

Comment (vassdeniss) [reply src/data/auth.js]: Ah, I never considered that as a possibility. Seems very reasonable now, and I will make a new commit as soon as I can.

**Review by viktorpts — CHANGES_REQUESTED** [review]
Having two places which individually call `clearUserData` is quite redundant, especially now that you're introduced a try-catch block. Would it not be more concise to have the try-catch inside the `logout` function, which executes `clearUserData` inside the `finally` block, so that it's cleared regardless of the response?

Alternatively, you can make both functions synchronous, with `logoutView` calling `logout` and redirecting without waiting for response and `logout` either storing the token in a variable, so that the storage can be cleared **before** sending both requests, or using `then()` to handle the requests in a callback while unconditionally clearing the storage.

**Comment by vassdeniss** [comment]
Clearing before the requests would be impossible as of now because the `api` will try to get the token for an authenticated request. From my testing I also noticed that a clear `finally` block results in a bit of a bad UX because *if* the requests fail an error would be thrown resulting in a `alert` from the `handleError` in the `api` which **will clear** the storage but the user would be stuck on the same page because the `throw` would make the `ctx` redirect to not execute.

Of course it is best to keep the storage operations in the `auth` and the redirect in the `logout`. Realistically the `catch` would be obsolete I believe? After all the `api` shows the error in an `alert` and we are not showing the error in the UI, we are only using the `finally` block for unconditional clearing if I'm thinking right.

I was thinking maybe 2 try blocks as in the first we can clear the storage regardless and in the second we can redirect regardless but I don't know how much of a fan I am of 2 try blocks. What would your suggestion be based on my observations?

**Comment by vassdeniss** [comment]
Another possibility is just making the session token an optional paramter in the `api` itself

**Comment by vassdeniss** [comment]
I went ahead and still did my double try block idea to see how it would work, and indeed it works as wanted. The only thing that I feel I did wrong is the error handling part. Up until now every error we handled on our own was meant to be seen by the user, if unhandled it would just pop an `alert` with the message. I am really not sure as to if we should have that `catch` block to begin with as the `api` handles `alerts` anyways, but really the only way to execute the clear is with the `finally` block. I am not certain of the practical correctness of a try-finally. I am curious and will wait for your feedback.

P.s. never expected a logout to become this complex, but at least I'm getting some learning learning.

**Review by viktorpts — CHANGES_REQUESTED** [review]
It looks much better, but note that a `try-finally` block does **not** catch errors! (which might be hinted by the omission of the `catch` block, but it's not really obvious). This is why you still need the second `try-finally` in `logoutView`. It would be better to include a `catch` block in `logout` that simply sets the exception as `handled`, so that an alert is not displayed by the API code. We don't care what happens with the request, so long as we are sure everything has been sent and there is no need for the storage anymore, and thus there is no need to inform the user of anything wrong which may have happened. That way you can omit the `try-finally` in `logoutView` and just have `await logout` directly followed by the redirect.

**Comment by vassdeniss** [comment]
Yes I understand it, I was really hesitant as to how to actually handle errors in this specific case.

**Review by viktorpts — APPROVED** [review]

**Pull request merged**